### PR TITLE
feat: init client protocol reference

### PR DESCRIPTION
**Pull request by ernado**
~~I can't understand how to build it (`yarn start` fails), can someone please help?
I think that some content from main repo should be copied to work.~~

I've added "Client protocol" category to "Reference / Development".
Most docs are from [go-faster.org/docs/clickhouse](https://go-faster.org/docs/clickhouse) and written by me while working on Go client [ch-go](https://github.com/ClickHouse/ch-go), so examples and reference are in Go.

Protocol reference is still incomplete, but should be good start.

### CODE REVIEW

**Comment by DanRoscigno** [comment]
Hi @ernado I will post instructions Monday, yes the docs from the main repo need to be copied in place.

**Comment by DanRoscigno** [comment]
@ernado this has not been merged yet, but have a look, it gives the details on building the docs locally with Docusaurus:

https://github.com/ClickHouse/ClickHouse/pull/39338/files

**Comment by ernado** [comment]
cc @gingerwizard, now this should work (thanks @DanRoscigno!)

**Comment by rfraposa** [comment]
This is great - thanks @ernado for creating this content. Dan - can we preview it in Vercel?

**Comment by DanRoscigno** [comment]
@ernado I missed that you had added it under reference and I put it in Integrations.  I will move it back!

File: docs/en/client-protocol/client.md [hunk 1]
```diff
@@ -0,0 +1,127 @@
+---
```
Comment (gingerwizard) [inline docs/en/client-protocol/client.md:1]: General comment - can we call this "native protocol "not "client protocol"

Comment (gingerwizard) [reply docs/en/client-protocol/client.md:1]: path should reflect this as well

File: docs/en/client-protocol/columns.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+:::tip
+
+Numeric types encoding matches memory layout of little endian cpus like AMD64 or ARM64.
+
+This allows to implement very efficient encoding and decoding.
+
+:::
+
+### Integers
+String of Int and UInt of 8, 16, 32, 64, 128 or 256 bits, in little endian.
```
Comment (gingerwizard) [inline docs/en/client-protocol/columns.md:20]: nit: new line

File: docs/en/client-protocol/columns.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+### Floats
+Float32 and Float64 in IEEE 754 binary representation.
```
Comment (gingerwizard) [inline docs/en/client-protocol/columns.md:23]: nit: new line

File: docs/en/client-protocol/columns.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+## String
+Just an array of String, i.e. (len, value).
```
Comment (gingerwizard) [inline docs/en/client-protocol/columns.md:26]: nit: new line

File: docs/en/client-protocol/columns.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+## FixedString(N)
+An array of N-byte sequences.
```
Comment (gingerwizard) [inline docs/en/client-protocol/columns.md:29]: nit: new line

File: docs/en/client-protocol/hash.md [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+ClickHouse uses **one of previous** versions of [CityHash from Google](https://github.com/google/cityhash).
+
+:::info
+CityHash has changed the algorithm after we have added it into ClickHouse.
+
+CityHash documentation specifically notes that the user should not rely to specific hash values and should not save it anywhere or use it as sharding key.
+
+But as we exposed this function to the user, we had to fix the version of CityHash (to 1.0.2). And now we guarantee that the behaviour of CityHash functions available in SQL will not change.
+
+— Alexey Milovidov, [comment](https://github.com/ClickHouse/ClickHouse/issues/8354#issuecomment-568517969) on issue.
```
Comment (gingerwizard) [inline docs/en/client-protocol/hash.md:20]: note sure we need to reference issues.

File: docs/en/client-protocol/hash.md [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+Don't use `farmHash64` to get Google's CityHash value! [FarmHash](https://opensource.googleblog.com/2014/03/introducing-farmhash.html) is a successor to CityHash, but they are not fully compatible.
+
+| String                                                     | ClickHouse64         | CityHash64          | FarmHash64           |
+|------------------------------------------------------------|----------------------|---------------------|----------------------|
+| `Moscow`                                                   | 12507901496292878638 | 5992710078453357409 | 5992710078453357409  |
+| `How can you write a big system without C++?  -Paul Glick` | 6237945311650045625  | 749291162957442504  | 11716470977470720228 |
+
+:::
+
+You can use [go-faster/city](https://github.com/go-faster/city) Go package that implements both variants.
```
Comment (gingerwizard) [inline docs/en/client-protocol/hash.md:36]: Add header "Language Implementations"

File: docs/en/client-protocol/hash.md [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+Also see [Introducing CityHash](https://opensource.googleblog.com/2011/04/introducing-cityhash.html) for description and
+reasoning behind creation. TL;DR **non-cryptographic** hash that is faster than [MurmurHash](http://en.wikipedia.org/wiki/MurmurHash), but more complex.
+
+## Examples
```
Comment (gingerwizard) [inline docs/en/client-protocol/hash.md:41]: I dont think we need these in the main docs

**Review by gingerwizard — CHANGES_REQUESTED** [review]
Thanks @ernado as ever

File: docs/en/client-protocol/basics.md [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+This document describes binary protocol for ClickHouse TCP clients.
+
+## Varint
+
+For lengths, packet codes and other cases the *unsigned varint* encoding is used.
+Use [binary.PutUvarint](https://pkg.go.dev/encoding/binary#PutUvarint) and [binary.ReadUvarint](https://pkg.go.dev/encoding/binary#ReadUvarint).
```
[the remark this answers is not in the document]
Comment (ernado) [reply docs/en/client-protocol/basics.md:21]: Agree, we can do it with docusaurus "tabs", hope somebody will contribute some python/java examples

File: docs/en/client-protocol/client.md (hunk 1 above)
Comment (ernado) [reply docs/en/client-protocol/client.md:1]: So no new naming for Native protocol, let's stick with native?

Comment (gingerwizard) [reply docs/en/client-protocol/client.md:1]: So lots of terms floating about:

1. Native protocol
2. Native format - supported over http
3. Native API - clickhouse-go uses this and it means "ClickHouse API" vs "sql/db" API - we need to rename.

In this case, these docs are for the native protocol and native format. I think "native protocol" is fine - although strictly it supports multiple formats e.g. RowBinary

**Comment by ernado** [comment]
I've updated according to comments. Not sure about cityhash, just delete this page if not needed, I'll keep it in go-faster docs.

File: docs/en/native-protocol/hash.md [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+Also see [Introducing CityHash](https://opensource.googleblog.com/2011/04/introducing-cityhash.html) for description and
+reasoning behind creation. TL;DR **non-cryptographic** hash that is faster than [MurmurHash](http://en.wikipedia.org/wiki/MurmurHash), but more complex.
+
+## Implementations
+
+### Go
+
+You can use [go-faster/city](https://github.com/go-faster/city) Go package that implements both variants.
+
+#### Benchmarks
```
Comment (gingerwizard) [inline docs/en/native-protocol/hash.md:41]: Lets drop these and examples for city hash - it remains a go-faster lib and I think that's how it will remain

Comment (ernado) [reply docs/en/native-protocol/hash.md]: Done

**Pull request merged**